### PR TITLE
Improve helper type operators through TypeScript 2.8

### PR DESCRIPTION
**Pull request by miZyind**
Mentioned in issue #1683

### CODE REVIEW

**Comment by clayne11** [comment]
This looks great. Can we get this reviewed?

File: typings/styled-components.d.ts [hunk 1]
```diff
@@ -54,7 +54,7 @@ type ThemedStyledComponentFactories<T> = {
 export interface ThemedBaseStyledInterface<T> extends ThemedStyledComponentFactories<T> {
   <P, TTag extends keyof JSX.IntrinsicElements>(tag: TTag): ThemedStyledFunction<P, T, P & JSX.IntrinsicElements[TTag]>;
   <P, O>(component: StyledComponentClass<P, T, O>): ThemedStyledFunction<P, T, O>;
-  <P extends { [prop: string]: any; theme?: T }>(component: React.ComponentType<P>): ThemedStyledFunction<P, T, WithOptionalTheme<P, T>>;
+  <P>(component: React.ComponentType<P>): T extends undefined ? StyledFunction<P> : ThemedStyledFunction<P, T, WithOptionalTheme<P, T>>;
```
Comment (Igorbek) [inline typings/styled-components.d.ts]: I think a cleaner return type would be `T extends { theme: T; } ? ThemedStyledFunction<P, T, WithOptionalTheme<P, T>> : ThemedStyledFunction<P, T>`

File: typings/styled-components.d.ts [hunk 2]
```diff
@@ -72,9 +72,7 @@ export interface ThemedCssFunction<T> {
 }
 
 // Helper type operators
-type Diff<T extends string, U extends string> = ({ [P in T]: P } & { [P in U]: never } & { [x: string]: never })[T];
-type Omit<T, K extends keyof T> = Pick<T, Diff<keyof T, K>>;
-type WithOptionalTheme<P extends { theme?: T; }, T> = Omit<P, "theme"> & { theme?: T; };
+type WithOptionalTheme<P extends { theme?: T; }, T> = Pick<P, Exclude<keyof P, "theme">> & { theme?: T; };
```
Comment (Igorbek) [inline typings/styled-components.d.ts]: if it's not going to be used anywhere else, then we can have only constraint `P extends { theme: T; }` (not optional)

**Comment by Igorbek** [comment]
@miZyind I'm actually not sure if my suggestions will work. Try them if you don't mind to see if it is better.

**Comment by miZyind** [comment]
@Igorbek  Sure! I'll try it later. Thanks for your advice.

**Comment by miZyind** [comment]
I've tried this:
```typescript
// Line 54~ in typings/styled-components.d.ts
export interface ThemedBaseStyledInterface<T> extends ThemedStyledComponentFactories<T> {
  // ...
  // <P>(component: React.ComponentType<P>): T extends undefined ? StyledFunction<P> : ThemedStyledFunction<P, T, WithOptionalTheme<P, T>>;
  <P>(component: React.ComponentType<P>): T extends { theme: T; } ? ThemedStyledFunction<P, T, WithOptionalTheme<P, T>> : ThemedStyledFunction<P, T>;
}
```

But the test case **issue1068.tsx** won't pass:
```typescript
// Line 7 ~ 9 in typings/tests/themed-tests/issue1068.tsx
interface ThemedOptionalProps extends OptionalProps {
  theme: MyTheme; // <- theme is necessary prop here
}

// Line 36 ~ 48 in typings/tests/themed-tests/issue1068.tsx
function optionalPropsWithRequiredTheme() {
  const Component = (props: ThemedOptionalProps) => <div>{props.text}</div>;

  const StyledComponent = styled(Component) ``;
  const StyledStyledComponent = styled(StyledComponent) ``;

  <Component theme={theme} />; // theme is required
  <Component text="test" theme={theme} />; // theme is required
  // Error: Property 'theme' is missing in type '{ text: string; }'.
  <StyledComponent text="test" />; // theme is optional
  <StyledComponent text="test" theme={theme} />; // theme is allowed
  <StyledStyledComponent text="test" />; // theme is optional
  // Error: Property 'theme' is missing in type '{ text: string; }'.
  <StyledStyledComponent text="test" theme={theme} />; // theme is allowed
}
```

That's because the type of this **StyledComponent** now inferred to:
```typescript
React.Component<{
  text?: string;
  theme: MyTheme
}>
```

instead of:
```typescript
React.Component<{
  text?: string;
  theme?: MyTheme
}>
```

I have no idea how to make this better :(

But this works fine(not optional theme prop in WithOptionalTheme):
```typescript
// Line 75 ~ 76 in typings/styled-components.d.ts
// Helper type operators
type WithOptionalTheme<P extends { theme: T; }, T> = Pick<P, Exclude<keyof P, "theme">> & { theme?: T; };
```

**Comment by Igorbek** [comment]
Let me look at this. I'll get back shortly.

**Comment by Igorbek** [comment]
@miZyind I tried another way, see commit: https://github.com/Igorbek/styled-components/commit/0b54a58be9fc64c7c2e85d699ab7337940604191
That makes `WithOptionalTheme` a conditional type that only removes `theme` if it was required in the original props.
It seems to work fine. Please take it for your PR.

**Comment by probablyup** [comment]
@Igorbek is this considered safe for a minor release, or should it wait until 4.0?

**Comment by Igorbek** [comment]
For TypeScript users it is definitely a breaking change as it now requires a higher version (2.8).
To me it seems that should be ok to go in minor release version with documenting that change is breaking.

**Comment by probablyup** [comment]
@kitten what's your opinion? If it's a breaking change, my inclination is it should wait for 4.0.

**Comment by clayne11** [comment]
I think it's OK to introduce new minor TS versions in a minor version bump. It's only a minor TS version change.

**Comment by probablyup** [comment]
Hey folks - so we've decided to [move the TS typings out of this repo](https://github.com/styled-components/styled-components/issues/1778) to DefinitelyTyped for 4.0. Please feel free to copy over the existing typings along with this change there!

**Pull request closed without merging**